### PR TITLE
Convert QoS APIs to pybind11

### PR DESCRIPTION
**Pull request by hidmic**
Part of #665. It's a shallow port (i.e. didn't migrate relevant `common.c` functions) in hopes we can eventually bind the entire `qos.py` module.

CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14064/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8884/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11756/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14155/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/14156/)

### CODE REVIEW

File: rclpy/src/rclpy/qos.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  int qos_liveliness,
+  py::capsule pyqos_liveliness_lease_duration,
+  bool avoid_ros_namespace_conventions)
+{
+  auto qos_profile =
+    std::unique_ptr<rmw_qos_profile_t, decltype(&PyMem_Free)>(
+    static_cast<rmw_qos_profile_t *>(
+      PyMem_Malloc(sizeof(rmw_qos_profile_t))), PyMem_Free);
+  if (!qos_profile) {
+    throw std::bad_alloc();
```
Comment (gbalke) [inline rclpy/src/rclpy/qos.cpp:66]: I believe this isn't documented in the header.

File: rclpy/src/rclpy/qos.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+namespace rclpy
+{
+
+namespace
+{
+
+// Fill a given rmw_time_t from an rclpy.duration.Duration instance.
+void
+_convert_py_duration_to_rmw_time(py::capsule pyduration, rmw_time_t * out_time)
```
Comment (gbalke) [inline rclpy/src/rclpy/qos.cpp:39]: This not being a bool definitely made the code below a lot simpler!

**Review by gbalke — APPROVED** [review]
Looking good!

File: rclpy/src/rclpy/qos.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+{
+  assert(out_time != nullptr);
+  auto duration = static_cast<rcl_duration_t *>(pyduration);
```
Comment (sloretz) [inline rclpy/src/rclpy/qos.cpp:42]: I'd recommend doing a `std::strcmp` on the capsule name to make sure this is a duration capsule.

Example: https://github.com/ros2/rclpy/blob/8a712bda9ebf1e7c5b3c94180584e6d687d9725d/rclpy/src/rclpy/client.cpp#L76-L79

File: rclpy/src/rclpy/qos.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    qos_profile.release(), "rmw_qos_profile_t",
+    [](PyObject * pycapsule_c) {
+      PyMem_Free(PyCapsule_GetPointer(pycapsule_c, "rmw_qos_profile_t"));
+    });
+}
+
+py::dict
+convert_to_py_qos_policy(py::capsule pyqos_profile)
+{
+  auto qos_profile = static_cast<rmw_qos_profile_t *>(pyqos_profile);
```
Comment (sloretz) [inline rclpy/src/rclpy/qos.cpp:98]: Also recommend checking the capsule name here

**Review by sloretz — APPROVED** [review]
Just a suggestion to check the pycapsule names.

File: rclpy/src/rclpy/qos.cpp (hunk 1 above)
Comment (hidmic) [reply rclpy/src/rclpy/qos.cpp:69]: Good point. See https://github.com/ros2/rclpy/pull/736/commits/e0d364465df44d78bd4fade06e65bb1be177467f.

File: rclpy/src/rclpy/qos.cpp (hunk 3 above)
Comment (hidmic) [reply rclpy/src/rclpy/qos.cpp:45]: Ah good catch! See https://github.com/ros2/rclpy/pull/736/commits/e0d364465df44d78bd4fade06e65bb1be177467f.

File: rclpy/src/rclpy/qos.cpp (hunk 4 above)
Comment (hidmic) [reply rclpy/src/rclpy/qos.cpp:104]: See https://github.com/ros2/rclpy/pull/736/commits/e0d364465df44d78bd4fade06e65bb1be177467f.

**Comment by hidmic** [comment]
CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14081/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8899/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11774/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14173/)

**Review by sloretz — APPROVED** [review]

**Comment by hidmic** [comment]
CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14087/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8905/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11780/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14181/)

**Comment by hidmic** [comment]
Alright, going in !

**Pull request merged**